nilearn/glm/_base.py

```diff
@@ -449,8 +476,223 @@ def _generate_filenames_output(
+            fields["entities"] = {"run": i_run + 1}  # type: ignore[assignment]
+        if generate_bids_name:
+            fields["entities"] = deepcopy(
+                model._reporting_data["run_imgs"][i_run]
+            )
+
+        tmp = {}
+        for extension in [FIGURE_FORMAT, "tsv"]:
+            for key, suffix in zip(
+                ["design_matrix", "correlation_matrix"],
+                ["design", "corrdesign"],
+            ):
+                fields["extension"] = extension
+                fields["suffix"] = suffix
+                tmp[f"{key}_{extension}"] = create_bids_filename(
+                    fields, entities_to_include
+                )
+
+        design_matrices_dict[i_run] = tempita.bunch(**tmp)
+
+    return design_matrices_dict
+
+
+def _generate_contrasts_dict(
+    model,
+    prefix: str,
+    contrasts,
+    design_matrices,
+    generate_bids_name: bool,
+    entities,
+    entities_to_include: list[str],
+) -> dict[int, dict[str, str]]:
+    """Return dictionary with filenames for contrast matrices figures.
+
+    contrasts_dict[i_run][contrast_name] = filename
+    """
+    fields = {
+        "prefix": prefix,
+        "extension": FIGURE_FORMAT,
+        "entities": deepcopy(entities),
+        "suffix": "design",
+    }
+    if generate_bids_name:
+        fields["prefix"] = None
+
+    contrasts_dict = tempita.bunch()
+
+    for i_run, _ in enumerate(design_matrices):
+        if _is_flm_with_single_run(model):
+            fields["entities"]["run"] = i_run + 1
+        if generate_bids_name:
+            fields["entities"] = deepcopy(
+                model._reporting_data["run_imgs"][i_run]
+            )
+
+        tmp = {}
+        for contrast_name in contrasts:
+            fields["entities"]["contrast"] = _clean_contrast_name(
+                contrast_name
+            )
+            tmp[contrast_name] = create_bids_filename(
+                fields, entities_to_include
+            )
+
+        contrasts_dict[i_run] = tempita.bunch(**tmp)
+
+    return contrasts_dict
+
+
+def _use_input_files_for_filenaming(self, prefix):
+    # for first level models with input files
+    # try to generate bids like output based on GLM input
+
+    if self.__str__() == "Second Level Model" or prefix is not None:
+        return False
+
+    input_files = self._reporting_data["run_imgs"]
+
+    files_used_as_input = all(len(x) > 0 for x in input_files.values())
+    tmp = {x.get("sub") for x in input_files.values()}
+    all_files_have_same_sub = len(tmp) == 1 and tmp is not None
+
+    return files_used_as_input and all_files_have_same_sub
+
+
+def _is_flm_with_single_run(model):
+    return (
+        model.__str__() == "First Level Model"
+        and len(model._reporting_data["run_imgs"]) > 1
+    )
+
+
 def _clean_contrast_name(contrast_name):
-    """Remove prohibited characters from name and convert to camelCase.
+    """Remove prohibited characters from name camelCase.
```

Comment:
The updated one is actually less clear to me

Reply:
I agree.